tests/unit/tf/outputs/test_contrastive.py

```diff
@@ -0,0 +1,219 @@
+    schema = ecommerce_data.schema
+    user_id = schema.select_by_tag(Tags.USER_ID)
+    item_id = schema.select_by_tag(Tags.ITEM_ID)
+
+    # TODO: Change this for new RetrievalModel
+    encoders = mm.SequentialBlock(
+        mm.ParallelBlock(
+            mm.EmbeddingTable(64, user_id.first), mm.EmbeddingTable(64, item_id.first)
+        ),
+        Rename(dict(user_id="query", item_id="candidate")),
+    )
+
+    mf = mm.Model(encoders, mm.ContrastiveOutput(item_id, "in-batch"))
+
+    testing_utils.model_test(mf, ecommerce_data, run_eagerly=True)
+
+
+def test_constrastive_mf_weights_in_output(ecommerce_data: Dataset):
+    schema = ecommerce_data.schema
+    schema["item_id"] = schema["item_id"].with_tags([Tags.TARGET])
+    user_id = schema.select_by_tag(Tags.USER_ID)
+    item_id = schema.select_by_tag(Tags.ITEM_ID)
+
+    # TODO: Change this for new RetrievalModel
+    encoder = mm.TabularBlock(mm.EmbeddingTable(64, user_id.first), aggregation="concat")
+
+    mf = mm.Model(encoder, mm.ContrastiveOutput(item_id, "in-batch"))
+
+    testing_utils.model_test(mf, ecommerce_data, run_eagerly=True)
+
+
+def test_two_tower_constrastive(ecommerce_data: Dataset):
+    model = mm.RetrievalModel(
+        mm.TwoTowerBlock(ecommerce_data.schema, query_tower=mm.MLPBlock([8])),
+        mm.ContrastiveOutput(
+            ecommerce_data.schema.select_by_tag(Tags.ITEM_ID),
+            negative_samplers="in-batch",
+            candidate_name="item",
+        ),
+    )
+
+    testing_utils.model_test(model, ecommerce_data)
+
+
+@pytest.mark.parametrize("run_eagerly", [True, False])
+def test_contrastive_output(ecommerce_data: Dataset, run_eagerly):
+    schema = ecommerce_data.schema
+    schema["item_category"] = schema["item_category"].with_tags(
+        schema["item_category"].tags + "target"
+    )
+    ecommerce_data.schema = schema
+    model = mm.Model(
+        mm.InputBlock(schema),
+        mm.MLPBlock([8]),
+        mm.ContrastiveOutput(
+            schema["item_category"],
+            negative_samplers=PopularityBasedSamplerV2(max_id=100, max_num_samples=20),
+        ),
+    )
+
+    _, history = testing_utils.model_test(model, ecommerce_data, run_eagerly=run_eagerly)
+
+    assert set(history.history.keys()) == {
+        "loss",
+        "ndcg_at_10",
+        "precision_at_10",
+        "map_at_10",
+        "mrr_at_10",
+        "recall_at_10",
+        "regularization_loss",
+    }
+
+
+def test_setting_negative_sampling_strategy(sequence_testing_data: Dataset):
+    dataloader, schema = _next_item_loader(sequence_testing_data, to_one_hot=False)
+    model_out = mm.ContrastiveOutput(schema["item_id_seq"], "in-batch")
+    model = mm.Model(mm.InputBlockV2(schema), mm.MLPBlock([32]), model_out)
+    model.compile(optimizer="adam")
+
+    batch = next(iter(dataloader))
+    output = model(batch[0], batch[1], training=True)
+    assert output[1].shape == (batch[1].shape[0], 51)
+
+    model_out.compile(
```

Comment:
Is this `compile` a similar compilation sense to the `model.compile`? It looks like this was already here on the block that has been rename. However, based on what it looks like it's doing, perhaps a different name could be picked if it's a way to change the samplers being used in this layer. (`update_negative_samplers` ?)